shim/src/realm.js

```diff
@@ -1,61 +1,226 @@
-import { createSandbox } from "./sandbox.js";
-import { sanitize } from "./sanitize.js";
+import { createSandbox, setSandboxGlobalObject } from "./sandbox.js";
 import { evaluate } from "./evaluate.js";
-import { getEvaluators } from "./evaluators.js";
 import { getStdLib } from "./stdlib.js";
 import { getIntrinsics } from "./intrinsics.js";
-import { proxyHandler } from "./proxy.js";
+import { assert, IsCallable } from "./utils";
 
-const RealmToSandbox = new WeakMap();
+const RealmRecord = Symbol('Realm Slot');
+const Intrinsics = Symbol('Intrinsics Slot');
+const GlobalObject = Symbol('GlobalObject Slot');
+const GlobalThisValue = Symbol('GlobalThisValue Slot');
+const GlobalEnv = Symbol('GlobalEnv Slot');
+const EvalHook = Symbol('EvalHook Slot');
+const IsDirectEvalHook = Symbol('IsDirectEvalHook Slot');
+const ImportHook = Symbol('ImportHook Slot');
+const ImportMetaHook = Symbol('ImportMetaHook Slot');
+const ShimSandbox = Symbol('Sandbox');
 
-function getSandbox(realm) {
-    const sandbox = RealmToSandbox.get(realm);
-    if (!sandbox) {
-        throw new Error(`Invalid realm object.`);
-    }
+// shim specific
+function getSandbox(realmRec) {
+    const sandbox = realmRec[ShimSandbox];
+    assert(typeof sandbox === 'object');
     return sandbox;
 }
 
+function getCurrentRealmRecord() {
+    let realmRec = window[RealmRecord];
+    if (!realmRec) {
+        // this is an outer realm, and we should set up the RealmRecord
+        window[RealmRecord] = {
+            // TODO: mimic what the global realm record should have
+            // including default hooks, etc.
+        };
+    }
+    return realmRec;
+}
+
+// <!-- es6num="8.1.2.5" -->
+function NewGlobalEnvironment(G, thisValue) {
+    // divering from spec to accomodate the iframe as the lexical environment
+    // using a class for better debugability
+    class EnvironmentRecord {
+        constructor(/*globalObject*/) {
+            this[GlobalThisValue] = thisValue;
+        }
+    }
+    return new EnvironmentRecord(G);
+}
+
+// <!-- es6num="8.2.3" -->
+function SetRealmGlobalObject(realmRec, globalObj, thisValue) {
+    if (globalObj === undefined) {
+        const intrinsics = realmRec[Intrinsics];
+        globalObj = Object.create(intrinsics['ObjectPrototype']);
+    }
+    assert(typeof globalObj === 'object');
+    if (thisValue === undefined) thisValue = globalObj;
```

Comment:
Yes, in most cases they will point to the same thing. We can discuss the details next Tue during the meeting or offline, but this is not really a new concept.